### PR TITLE
Implemented rating resources function (Fixes #269)

### LINKED ISSUE #269: Rating resources
Provide the method for user to provide rating about resource along with short comment (optional)
User should be able to update its rating anytime which should not enter as new one but update previous update for same resource.
Rating should be calculated on Resource page which seems to be static right now.

### PR DESCRIPTION
**Pull request by yaoxing1990**
Implemented UI for user to submit rating:
capture

### CODE REVIEW

**Comment by yaoxing1990** [comment]
Implemented UI with comment. Added updating to database feature. Every time the same user gives a new rating, the old rating will be updated.

**Comment by yaoxing1990** [comment]
I want to extract route parameter in ngOnInit but once I insert code into it, once restart, the app fails launching. Is there anything wrong with this? @paulbert @lmmrssa @nirojdyola 
```javascript
 this.route.paramMap.pipe(switchMap((params: ParamMap) => {
      this.getResource(params.get('id'));
      this.id = params.get('id');
    }))
 .subscribe(resource => this.resource = resource);
```

**Comment by yaoxing1990** [comment]
@lmmrssa @paulbert Finished implementing resource rating feature (#269)
For example, 1 female gave 2.6 points, 1 male gave 4.8 points, so there are 2 people rating, the average point is (2.6 * 1 + 4.8 * 1) / 2 = 3.7, 3.7 * 20% = 74%
capture2
capture1

**Comment by lmmrssa** [comment]
@yaoxing1990 It seems you have added package-lock.json file, could you remove that? Also you have changed some of the package versions and added ts-lint on production. Please revert that as well.

**Comment by yaoxing1990** [comment]
@lmmrssa Reverted

File: package.json [hunk 1]
```diff
@@ -58,7 +58,7 @@
     "protractor": "~5.1.2",
     "request-promise": "^4.2.1",
     "ts-node": "~3.2.0",
-    "tslint": "~5.3.2",
+    "tslint": "^5.3.2",
```
Comment (razu9861) [inline package.json]: why tslint is aded in package.json, its already been added you dont have to change it

File: src/app/resources/rate-resources/resources-rate.component.ts [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    FormArray,
+    Validators
+  } from '@angular/forms';
+
+@Component({
+  templateUrl: './resources-rate.component.html'
+})
+export class ResourcesRateComponent implements OnInit {
+
+  ratingForm: FormGroup;
```
Comment (razu9861) [inline src/app/resources/rate-resources/resources-rate.component.ts]: please follow tslint rule for indentation

File: package.json (hunk 1 above)
Comment (yaoxing1990) [reply package.json]: @razu9861 I accidentally changed it form "~" to "^"

File: src/app/resources/rate-resources/resources-rate.component.ts (hunk 2 above)
Comment (yaoxing1990) [reply src/app/resources/rate-resources/resources-rate.component.ts]: @razu9861 https://github.com/ole-vi/planet/pull/270/commits/06a0b4bef34cf7d1fb438d66817345ae3496b839

File: package.json (hunk 1 above)
Comment (nirojdyola) [reply package.json]: Yes, this change is not required for the given issue. The issue requires only rating function . So, you don;t have to change it. Could you revert it back ?

File: src/app/resources/rate-resources/resources-rate.component.html [hunk 3]
```diff
@@ -0,0 +1,45 @@
+<h1 i18n>Submit your rating</h1>
+<form [formGroup]="ratingForm" (ngSubmit)="onSubmit()">
+
```
Comment (nirojdyola) [inline src/app/resources/rate-resources/resources-rate.component.html]: Could you please remove extra line ?

File: src/app/resources/rate-resources/resources-rate.component.html [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    <div hidden>
+        <mat-form-field>
+            <input matInput i18n-placeholder [(ngModel)]="this.id" type="text" formControlName="id">
+            <mat-error><planet-form-error-messages [control]="ratingForm.controls.id"></planet-form-error-messages></mat-error>
+        </mat-form-field>
+    </div>
+
```
Comment (nirojdyola) [inline src/app/resources/rate-resources/resources-rate.component.html]: Remove extra line please.

File: src/app/resources/rate-resources/resources-rate.component.html [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    <div hidden>
+        <mat-form-field>
+            <input matInput i18n-placeholder [(ngModel)]="this.user" type="text" formControlName="user">
+            <mat-error><planet-form-error-messages [control]="ratingForm.controls.user"></planet-form-error-messages></mat-error>
+        </mat-form-field>
+    </div>
+
```
Comment (nirojdyola) [inline src/app/resources/rate-resources/resources-rate.component.html]: There are more extra line . Could you please remove all those ?

File: src/app/resources/resources.component.ts [hunk 6]
```diff
@@ -10,7 +10,10 @@ export class ResourcesComponent implements OnInit, AfterViewInit {
   resources = new MatTableDataSource();
   @ViewChild(MatPaginator) paginator: MatPaginator;
   displayedColumns = [ 'title', 'rating' ];
-  readonly dbName = 'resources';
+  readonly resourceDb = 'resources';
```
Comment (nirojdyola) [inline src/app/resources/resources.component.ts]: I don't understand why readonly dbName = 'resources'; has been changed to readonly resourceDb = 'resources';

File: package.json (hunk 1 above)
Comment (yaoxing1990) [reply package.json]: @nirojdyola Reverted

File: src/app/resources/rate-resources/resources-rate.component.html (hunk 5 above)
Comment (yaoxing1990) [reply src/app/resources/rate-resources/resources-rate.component.html]: @nirojdyola All removed https://github.com/ole-vi/planet/pull/270/commits/45ffaf3d071f9cb9676524bb17fa8f9484341306

File: src/app/resources/resources.component.ts (hunk 6 above)
Comment (yaoxing1990) [reply src/app/resources/resources.component.ts]: @nirojdyola

File: src/app/resources/rate-resources/resources-rate.component.html [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+    </div>
+    <div>
+        <mat-radio-group>
+            <label i18n>Gender:</label>
+            <mat-radio-button value="true" name="genderToggler" (change)="isMale(true)" i18n>Male</mat-radio-button>
+            <mat-radio-button value="false" name="genderToggler" (change)="isMale(false)" i18n>Female</mat-radio-button>
+          </mat-radio-group>
+    </div>
+    <div>
+        <textarea matInput i18n-placeholder placeholder="Type your rating: (0-5)" formControlName="rating"></textarea>
```
Comment (lmmrssa) [inline src/app/resources/rate-resources/resources-rate.component.html]: It should be some radio button or other input not plain text field

File: src/app/resources/rate-resources/resources-rate.component.html [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+    <div hidden>
+        <mat-form-field>
+            <input matInput i18n-placeholder [(ngModel)]="this.user" type="text" formControlName="user">
+            <mat-error><planet-form-error-messages [control]="ratingForm.controls.user"></planet-form-error-messages></mat-error>
+        </mat-form-field>
+    </div>
+    <div>
+        <mat-radio-group>
+            <label i18n>Gender:</label>
+            <mat-radio-button value="true" name="genderToggler" (change)="isMale(true)" i18n>Male</mat-radio-button>
```
Comment (lmmrssa) [inline src/app/resources/rate-resources/resources-rate.component.html]: this is not to be input by user instead it should be taken by default based on logged in user

File: src/app/resources/rate-resources/resources-rate.component.ts [hunk 9]
```diff
@@ -0,0 +1,126 @@
+import { switchMap } from 'rxjs/operators';
+import { Component, OnInit } from '@angular/core';
+import { CouchService } from '../../shared/couchdb.service';
+import { UserService } from '../../shared/user.service';
+
+import { ActivatedRoute, ParamMap } from '@angular/router';
+import { DomSanitizer } from '@angular/platform-browser';
+import { Router } from '@angular/router';
```
Comment (lmmrssa) [inline src/app/resources/rate-resources/resources-rate.component.ts]: this can merge with line 6

Comment (yaoxing1990) [reply src/app/resources/rate-resources/resources-rate.component.ts]: @lmmrssa All fixed https://github.com/ole-vi/planet/pull/270/commits/9540e845997088564010b8e9996cb37f1b38fb71

**Comment by yaoxing1990** [comment]
@lmmrssa Because during registration, user gender is not recorded, so currently the gender cannot be obtained, Also the new database not in installation script is still a issue

**Comment by lmmrssa** [comment]
Error Received
```
 92% chunk asset optimization                                                 y <--- Last few GCs --->

  497502 ms: Mark-sweep 650.1 (1167.4) -> 449.3 (1143.4) MB, 3961.7 / 0.0 ms [allocation failure] [GC in old space requested].
  499260 ms: Mark-sweep 449.3 (1399.4) -> 449.1 (1343.4) MB, 1756.6 / 0.0 ms [allocation failure] [GC in old space requested].
  500839 ms: Mark-sweep 449.1 (1599.4) -> 446.9 (1450.4) MB, 1577.0 / 0.0 ms [last resort gc].
  502102 ms: Mark-sweep 446.9 (1450.4) -> 446.9 (1419.4) MB, 1262.9 / 0.0 ms [last resort gc].


<--- JS stacktrace --->

==== JS stack trace =========================================

Security context: 0x2e04a07cf781 <JS Object>
    1: flatten_args(aka flatten_args) [0x2e04a0704381 <undefined>:~11681] [pc=0xc7593c3eb59] (this=0x2e04a0704381 <undefined>,fn=0x358135d93731 <an AST_Function with map 0x112ae206ce79>,scope=0x358135d95499 <an AST_Function with map 0x196f7e2b9a59>)
    2: /* anonymous */(aka /* anonymous */) [0x2e04a0704381 <undefined>:11627] [pc=0xc7593dd885f] (this=0x2e04a0704381 <undefined>,self=0x2e2f9f6e0...

FATAL ERROR: Committing semi space failed. Allocation failed - process out of memory
 1: node::Abort() [ng]
 2: 0x109f13c [ng]
 3: v8::Utils::ReportApiFailure(char const*, char const*) [ng]
 4: v8::internal::V8::FatalProcessOutOfMemory(char const*, bool) [ng]
 5: v8::internal::Heap::PerformGarbageCollection(v8::internal::GarbageCollector, v8::GCCallbackFlags) [ng]
 6: v8::internal::Heap::CollectGarbage(v8::internal::GarbageCollector, char const*, char const*, v8::GCCallbackFlags) [ng]
 7: v8::internal::Factory::NewRawOneByteString(int, v8::internal::PretenureFlag) [ng]
 8: v8::internal::Factory::NewStringFromOneByte(v8::internal::Vector<unsigned char const>, v8::internal::PretenureFlag) [ng]
 9: v8::internal::Factory::NumberToString(v8::internal::Handle<v8::internal::Object>, bool) [ng]
10: v8::internal::Object::ToString(v8::internal::Isolate*, v8::internal::Handle<v8::internal::Object>) [ng]
11: v8::internal::Object::ConvertToName(v8::internal::Isolate*, v8::internal::Handle<v8::internal::Object>) [ng]
12: v8::internal::LookupIterator::PropertyOrElement(v8::internal::Isolate*, v8::internal::Handle<v8::internal::Object>, v8::internal::Handle<v8::internal::Object>, bool*, v8::internal::LookupIterator::Configuration) [ng]
13: v8::internal::Runtime::GetObjectProperty(v8::internal::Isolate*, v8::internal::Handle<v8::internal::Object>, v8::internal::Handle<v8::internal::Object>) [ng]
14: 0xed4961 [ng]
15: v8::internal::Runtime_KeyedGetProperty(int, v8::internal::Object**, v8::internal::Isolate*) [ng]
16: 0xc758e8092a7
Aborted
```

**Comment by yaoxing1990** [comment]
@paulbert No problem. Is it caused by rebase? Laxman did rebase on this branch last night

**Comment by paulbert** [comment]
Travis failing because it's taking too long to build...
image

**Comment by razu9861** [comment]
@yaoxing1990  i have attatched images for rating please implement this, have a look at both images add a button for rating on second image while clicking buton it should show popup for ratng. please try to implement it 
dashboard-v2v2 42-lib-layout-2b
screen shot 2018-01-11 at 10 01 33 am

**Comment by paulbert** [comment]
Closing since this was incorporated with #321

**Pull request closed without merging**